File: server/src/modules/okr.js
import { z } from 'zod';
import { round, clamp, band } from '../utils/scoring.js';

/* ============================================================
   MODULE 10 — OKR (OBJECTIVES AND KEY RESULTS)
   Goal-setting and tracking framework. Sets the framework for
   each objective and tracks achievement, so the customer
   actually finishes what they started.
   ============================================================ */

export const key = 'okr';
export const title = 'Objectives & Key Results';
export const action = 'MAPPING';

export const KR_STATUSES = ['NOT_STARTED', 'ON_TRACK', 'AT_RISK', 'OFF_TRACK', 'ACHIEVED'];

const keyResultSchema = z.object({
  id: z.string().optional(),
  description: z.string().min(3),
  metric: z.string().min(1),
  baseline: z.number().default(0),
  target: z.number(),
  current: z.number().default(0),
  unit: z.string().default(''),
  dueDate: z.coerce.date().optional(),
});

const objectiveSchema = z.object({
  id: z.string().optional(),
  title: z.string().min(3),
  owner: z.string().optional(),
  quarter: z.string().optional(),
  keyResults: z.array(keyResultSchema).min(1, 'An objective needs at least one key result'),
});

export const inputSchema = z.object({
  objectives: z.array(objectiveSchema).min(1, 'Supply at least one objective'),
  /** Enforce that objectives are closed out, not abandoned. */
  enforceCompletion: z.boolean().default(true),
  asOf: z.coerce.date().optional(),
});

export function run(input) {
  const data = inputSchema.parse(input);
  const asOf = data.asOf ?? new Date();

  const objectives = data.objectives.map((objective, oi) => {
    const keyResults = objective.keyResults.map((kr, ki) => scoreKeyResult(kr, ki, asOf));
    // Objective progress is the mean of its key results — standard OKR practice.
    const progress = round(keyResults.reduce((sum, kr) => sum + kr.progress, 0) / keyResults.length, 1);

    return {
      id: objective.id ?? `obj-${oi + 1}`,
      title: objective.title,
      owner: objective.owner ?? null,
      quarter: objective.quarter ?? null,
      progress,
      status: statusForProgress(progress, keyResults),
      achieved: progress >= 100,
      keyResults,
      atRiskCount: keyResults.filter((kr) => kr.status === 'AT_RISK' || kr.status === 'OFF_TRACK').length,
    };
  });

  const overallProgress = round(
    objectives.reduce((sum, o) => sum + o.progress, 0) / objectives.length,
    1
  );
  const achievedCount = objectives.filter((o) => o.achieved).length;

  // The score *is* the tracked progress — this module measures follow-through.
  const score = clamp(round(overallProgress));

  return {
    score,
    output: {
      asOf: asOf.toISOString(),
      framework: {
        objectiveCount: objectives.length,
        keyResultCount: objectives.reduce((n, o) => n + o.keyResults.length, 0),
        cadence: 'quarterly',
      },
      overallProgress,
      achievedObjectives: achievedCount,
      completionRate: round((achievedCount / objectives.length) * 100, 1),
      objectives,
      blockers: objectives.flatMap((o) =>
        o.keyResults
          .filter((kr) => kr.status === 'OFF_TRACK')
          .map((kr) => ({ objective: o.title, keyResult: kr.description, gap: kr.gap }))
      ),
      health: band(score),
      enforcement: data.enforceCompletion
        ? {
            enabled: true,
            unfinished: objectives.filter((o) => !o.achieved).map((o) => o.title),
            message:
              achievedCount === objectives.length
                ? 'All objectives achieved.'
                : `${objectives.length - achievedCount} objective(s) remain open — the cycle cannot be closed until they are achieved or formally dropped.`,
          }
        : { enabled: false },
      summary: `${overallProgress}% overall progress; ${achievedCount}/${objectives.length} objectives achieved.`,
      recommendations: buildRecommendations(objectives, overallProgress),
    },
  };
}

function scoreKeyResult(kr, index, asOf) {
  const span = kr.target - kr.baseline;
  // Guard the degenerate case where target === baseline.
  const progress = span === 0
    ? (kr.current >= kr.target ? 100 : 0)
    : clamp(round(((kr.current - kr.baseline) / span) * 100, 1));

  const overdue = kr.dueDate ? asOf > kr.dueDate && progress < 100 : false;

  let status;
  if (progress >= 100) status = 'ACHIEVED';
  else if (progress === 0) status = 'NOT_STARTED';
  else if (overdue || progress < 30) status = 'OFF_TRACK';
  else if (progress < 60) status = 'AT_RISK';
  else status = 'ON_TRACK';

  return {
    id: kr.id ?? `kr-${index + 1}`,
    description: kr.description,
    metric: kr.metric,
    unit: kr.unit,
    baseline: kr.baseline,
    target: kr.target,
    current: kr.current,
    progress,
    gap: round(kr.target - kr.current, 2),
    dueDate: kr.dueDate ? kr.dueDate.toISOString() : null,
    overdue,
    status,
  };
}

function statusForProgress(progress, keyResults) {
  if (progress >= 100) return 'ACHIEVED';
  if (keyResults.some((kr) => kr.status === 'OFF_TRACK')) return 'OFF_TRACK';
  if (progress < 60) return 'AT_RISK';
  if (progress === 0) return 'NOT_STARTED';
  return 'ON_TRACK';
}

function buildRecommendations(objectives, overallProgress) {
  const out = [];
  const offTrack = objectives.filter((o) => o.status === 'OFF_TRACK');
  if (offTrack.length) out.push(`${offTrack.length} objective(s) off track: ${offTrack.map((o) => o.title).join(', ')}.`);
  if (overallProgress < 40) out.push('Overall progress below 40% — re-scope the objectives or reallocate ownership.');
  const unowned = objectives.filter((o) => !o.owner);
  if (unowned.length) out.push(`${unowned.length} objective(s) have no owner — unowned OKRs do not get finished.`);
  if (!out.length) out.push('OKR cycle is healthy — maintain the weekly check-in cadence.');
  return out;
}

export default { key, title, action, inputSchema, run };

File: server_python/modules/okr.py
"""MODULE 10 — OKR / OBJECTIVES AND KEY RESULTS
(port of server/src/modules/okr.js)
Goal-setting and tracking framework. Sets the framework for each objective
and tracks achievement, so the customer actually finishes what they started.
"""
from datetime import datetime, timezone
from typing import List, Optional
from pydantic import BaseModel, Field

from scoring import js_round, clamp, band

KEY = 'okr'
TITLE = 'Objectives & Key Results'
ACTION = 'MAPPING'

KR_STATUSES = ['NOT_STARTED', 'ON_TRACK', 'AT_RISK', 'OFF_TRACK', 'ACHIEVED']


class KeyResult(BaseModel):
    id: Optional[str] = None
    description: str = Field(min_length=3)
    metric: str = Field(min_length=1)
    baseline: float = 0
    target: float
    current: float = 0
    unit: str = ''
    dueDate: Optional[datetime] = None


class Objective(BaseModel):
    id: Optional[str] = None
    title: str = Field(min_length=3)
    owner: Optional[str] = None
    quarter: Optional[str] = None
    keyResults: List[KeyResult] = Field(min_length=1, description='An objective needs at least one key result')


class InputSchema(BaseModel):
    objectives: List[Objective] = Field(min_length=1, description='Supply at least one objective')
    # Enforce that objectives are closed out, not abandoned.
    enforceCompletion: bool = True
    asOf: Optional[datetime] = None


def _iso(dt):
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc).isoformat().replace('+00:00', 'Z')


def run(raw_input, context=None):
    data = InputSchema.model_validate(raw_input or {})
    as_of = data.asOf or datetime.now(timezone.utc)
    if as_of.tzinfo is None:
        as_of = as_of.replace(tzinfo=timezone.utc)

    objectives = []
    for oi, objective in enumerate(data.objectives):
        key_results = [_score_key_result(kr, ki, as_of) for ki, kr in enumerate(objective.keyResults)]
        # Objective progress is the mean of its key results — standard OKR practice.
        progress = js_round(sum(kr['progress'] for kr in key_results) / len(key_results), 1)

        objectives.append({
            'id': objective.id or f'obj-{oi + 1}',
            'title': objective.title,
            'owner': objective.owner,
            'quarter': objective.quarter,
            'progress': progress,
            'status': _status_for_progress(progress, key_results),
            'achieved': progress >= 100,
            'keyResults': key_results,
            'atRiskCount': len([kr for kr in key_results if kr['status'] in ('AT_RISK', 'OFF_TRACK')]),
        })

    overall_progress = js_round(sum(o['progress'] for o in objectives) / len(objectives), 1)
    achieved_count = len([o for o in objectives if o['achieved']])

    # The score *is* the tracked progress — this module measures follow-through.
    score = clamp(js_round(overall_progress))

    if data.enforceCompletion:
        enforcement = {
            'enabled': True,
            'unfinished': [o['title'] for o in objectives if not o['achieved']],
            'message': (
                'All objectives achieved.'
                if achieved_count == len(objectives)
                else f'{len(objectives) - achieved_count} objective(s) remain open — the cycle cannot '
                     'be closed until they are achieved or formally dropped.'
            ),
        }
    else:
        enforcement = {'enabled': False}

    return {
        'score': score,
        'output': {
            'asOf': _iso(as_of),
            'framework': {
                'objectiveCount': len(objectives),
                'keyResultCount': sum(len(o['keyResults']) for o in objectives),
                'cadence': 'quarterly',
            },
            'overallProgress': overall_progress,
            'achievedObjectives': achieved_count,
            'completionRate': js_round((achieved_count / len(objectives)) * 100, 1),
            'objectives': objectives,
            'blockers': [
                {'objective': o['title'], 'keyResult': kr['description'], 'gap': kr['gap']}
                for o in objectives
                for kr in o['keyResults']
                if kr['status'] == 'OFF_TRACK'
            ],
            'health': band(score),
            'enforcement': enforcement,
            'summary': f'{overall_progress}% overall progress; {achieved_count}/{len(objectives)} objectives achieved.',
            'recommendations': _build_recommendations(objectives, overall_progress),
        },
    }


def _score_key_result(kr, index, as_of):
    span = kr.target - kr.baseline
    # Guard the degenerate case where target == baseline.
    if span == 0:
        progress = 100 if kr.current >= kr.target else 0
    else:
        progress = clamp(js_round(((kr.current - kr.baseline) / span) * 100, 1))

    due = kr.dueDate
    if due is not None and due.tzinfo is None:
        due = due.replace(tzinfo=timezone.utc)
    overdue = bool(due and as_of > due and progress < 100)

    if progress >= 100:
        status = 'ACHIEVED'
    elif progress == 0:
        status = 'NOT_STARTED'
    elif overdue or progress < 30:
        status = 'OFF_TRACK'
    elif progress < 60:
        status = 'AT_RISK'
    else:
        status = 'ON_TRACK'

    return {
        'id': kr.id or f'kr-{index + 1}',
        'description': kr.description,
        'metric': kr.metric,
        'unit': kr.unit,
        'baseline': kr.baseline,
        'target': kr.target,
        'current': kr.current,
        'progress': progress,
        'gap': js_round(kr.target - kr.current, 2),
        'dueDate': _iso(due) if due else None,
        'overdue': overdue,
        'status': status,
    }


def _status_for_progress(progress, key_results):
    if progress >= 100:
        return 'ACHIEVED'
    if any(kr['status'] == 'OFF_TRACK' for kr in key_results):
        return 'OFF_TRACK'
    if progress < 60:
        return 'AT_RISK'
    if progress == 0:
        return 'NOT_STARTED'
    return 'ON_TRACK'


def _build_recommendations(objectives, overall_progress):
    out = []
    off_track = [o for o in objectives if o['status'] == 'OFF_TRACK']
    if off_track:
        out.append(f'{len(off_track)} objective(s) off track: {", ".join(o["title"] for o in off_track)}.')
    if overall_progress < 40:
        out.append('Overall progress below 40% — re-scope the objectives or reallocate ownership.')
    unowned = [o for o in objectives if not o['owner']]
    if unowned:
        out.append(f'{len(unowned)} objective(s) have no owner — unowned OKRs do not get finished.')
    if not out:
        out.append('OKR cycle is healthy — maintain the weekly check-in cadence.')
    return out
